### DISCUSSION TITLE
Error: Cannot read properties of undefined (reading 'startsWith')

### DISCUSSION THREAD

**Discussion by chasm** [vercel/community#3986] [Help] [answered]
### Summary

I have set up an `api` folder at project root with a subfolder `contact` and a file `index.ts`. The contents of the file are copied and pasted directly from the Vercel site:

```ts
import type { VercelRequest, VercelResponse } from "@vercel/node"

export default function handler(
  request: VercelRequest,
  response: VercelResponse,
) {
  response.status(200).json({
    body: request.body,
    query: request.query,
    cookies: request.cookies,
  })
}
```

I have a simple form with action `/api/contact` and method `post`:

```html
<form
  name="contactForm"
  class="sb-form"
  action="/api/contact"
  id="_Ps3wMwTuF4YyJ87LThVGNP"
  method="post"
>
  <button class="sb-button sb-button-submit" type="submit">
    Send your message
  </button>
</form>
```

When I submit this form, I get the following error:

```
502: BAD_GATEWAY
Code: NO_RESPONSE_FROM_FUNCTION
ID: dev1::kl2dg-1694231108634-7c62744ae342

Check the logs in your terminal window to see the application error.
```

In my terminal window I see:

```
Error: Cannot read properties of undefined (reading 'startsWith')
```

Sadly, this error message is cryptic. I have searched **all** of my code *including node_modules* and I can find `startsWith` nowhere. Not one instance. Where on Earth is that coming from?

I have the following devDependencies (and no others):

```
"@astrojs/mdx": "^1.0.1",
"@astrojs/sitemap": "^3.0.0",
"@js-temporal/polyfill": "^0.4.4",
"@mailerlite/mailerlite-nodejs": "^1.1.0",
"@sitebender/components": "^0.1.2",
"@types/qrcode": "^1.5.2",
"@vitest/coverage-v8": "^0.34.3",
"astro": "^3.0.9",
"color-contrast-calc": "^0.4.1",
"dprint": "^0.40.2",
"exifr": "^7.1.3",
"exifreader": "^4.13.0",
"postmark": "^3.0.19",
"qrcode": "^1.5.3",
"rollup": "^3.29.0",
"typescript": "^5.2.2",
"vercel": "^32.2.0",
"vite": "^4.4.9",
"vite-plugin-compress": "^2.1.1",
"vitest": "^0.34.3"
```

I have searched the issues on GitHub, I have Googled the error, I can find nothing about it anywhere. I am at a complete loss and an utter standstill.

Ideas, anyone? Has anyone else gotten this error? Is there any way to get more information about the error?

Note: I am working locally with `vercel dev`. I am also using:

* node v20.6.1
* npm 9.8.1
* pnpm 8.7.4

And I'm using a MacBook Pro 15-inch 2019 2.3 GHz 8-Core Intel Core i9 with 32 GB 2400 MHz DDR4 running Ventura 13.5.2 (22G91) with the latest VSCode, iTerm2, and zsh.

TIA for any help anyone can offer. I'm out of ideas.

### Steps to Reproduce

I have provided everything I can think of above. I will try to reproduce this in an empty demo and will update this issue with the results.

**Comment by chasm** [comment, marked as answer]
In putting together a simple example, I got a different error. It became clear that the problem is node itself. I backed my node version down to 18 – **very** annoying as I had been on 19 until the recent upgrade and brew does not make it easy to go back – and it works now (after further fiddling with `tsconfig.json`).

For anyone else running on the edge and discovering this problem, sadly, this seems to be the only workaround. I tried using node HEAD, but no joy, and 20.0.6 doesn't fix it either.

I'll try to remember to update this if/when node (or astro or vercel) figures out the fix.

**Comment by yanggggjie** [comment]
Thanks for the tip. I tested this and it has nothing to do with the node version, but with tsconfig.json. 
This worked for me, 
1. Change "moduleResolution": "bundler" to "moduleResolution": "Node"
2. Remove "allowImportingTsExtensions": true.
😄

**Comment by augustl** [comment]
I got this error when I read an environment variable at build-time. that wasn't actually set in Vercel.
